src/elm/Api/Graphql.elm

```diff
@@ -7,15 +7,28 @@ import Graphql.SelectionSet exposing (SelectionSet)
 import Session.Shared exposing (Shared)
 
 
+withAuthToken : Maybe String -> Graphql.Http.Request decodesTo -> Graphql.Http.Request decodesTo
+withAuthToken authToken =
+    case authToken of
+        Just t ->
+            Graphql.Http.withHeader "authorization"
+                ("Bearer " ++ t)
+
+        Nothing ->
+            identity
+
+
 query : Shared -> SelectionSet a RootQuery -> (Result (Graphql.Http.Error a) a -> msg) -> Cmd msg
-query { endpoints } query_ toMsg =
+query { endpoints, authToken } query_ toMsg =
```

Comment:
maybe we can have two functions here: 

`publicQuery` and `query`. Because some queries won't have a token (like community query used for invitation) and others (the majority for sure) will have a token.

maybe:

```elm
query : LoggedIn.Model -> SelectionSet a RottQuery -> (. . .) -> Cmd msg
query { profile, shared } query_ toMsg =

publicQuery : Shared -> SelectionSet a  . . .
publicQuery {endpoints} query_ toMsg = 

```

This way the type signature already indicates us that some queries must be logged, while others don't.

Having the token saved to shared looks weird because we only got the token after logging in

Reply:
>maybe we can have two functions here

I tried that approach and I didn't like it. IMO the ideal signature would be:

```elm
query : Endpoint -> Maybe AuthToken -> SelectionSet a RootQuery -> ...
mutation : Endpoint -> Maybe AuthToken -> SelectionSet a RootQuery -> ...
```
`Maybe`s in the signature tell us everything we need to know and we don't have to support 2 functions with the same purpose. I tried a similar approach as well and I didn't like the increased complexity and the need to change the code in many places.

Also, we can't use `LoggedIn.Model` in `Api.Graphql` due to circular dependency, but we can achieve the same behavior with extensible records:

```elm
query : { loggedIn | shared : Shared, authToken : String } -> SelectionSet a RootQuery -> ...
```

I also tried it and didn't like it much... Although, it's nice not to have any logged-in related things in `Shared`, but we already have `Shared` "broken": there's `maybeAccount : (account, isPinAvailable)` in it. As an option, we can add the `authToken` as the third element to this tuple. I also tried it and I also didn't like it.

IMO, it's good enough to have `authToken` inside `Shared`. It allows us to keep all existing queries and mutations untouched: fewer changes, fewer errors.

Reply:
+1 on the new signature suggested, but -1 on having more logged-in related things because we already have it. Also that isPinAvailable is reminiscent from a time where our app supported only logging in when we needed something signed, very different from the approach we have now. I would dare to say this isPinAvailable always have the same value.

Maybe we can get into a middle ground here, keep your suggestion of ignoring `LoggedIn`/`Share` altogether and instead use `Endpoint` and `Maybe`, then send the `Just token` when we have it available on the caller method

Reply:
if we want to avoid keep dragging this issue longer because it would require bigger changes than we can ask for @NeoVier 's help here, this refactoring is quite big

Reply:
@lucca65 Yep, these changes will definitely take some time. If I could pass this task along and do more on Natus, I'll be really thankful 🙏

Reply:
I'm happy to help! Can I just get some context on what's left to do though?

Reply:
sure! This issue is the frontend implementation for https://github.com/cambiatus/backend/pull/126, that brings a basic API authentication flow to the app. Before this PR our API was public, but now certain queries will be scoped by the logged user and some mutations will only work if the user is changing its own data.

For example: we have an mutation that allows the user to delete their KYC data, before it would take one param `account` to know where it should delete the KYC data from, but now it has no arguments, taking instead the logged user. Another query is the `notifications` query, that took a `account` param but now has none.

The authentication flow is straight forward, whenever you call a `signUp` or `signIn` you will get a `Session` GraphQL object:

<img width="291" alt="Captura de Tela 2021-03-02 às 08 14 10" src="https://user-images.githubusercontent.com/1082127/109640466-48345280-7b2f-11eb-81f2-b7c43977d86a.png">
<img width="292" alt="Captura de Tela 2021-03-02 às 08 14 49" src="https://user-images.githubusercontent.com/1082127/109640532-5f734000-7b2f-11eb-9df3-144f702ed181.png">

Session contains a `token` that must be stored inside LoggedIn (or not, as you can see in this thread here). Only loggedIn users will have a `token`.

We have queries that don't require authentication such as the the `invite` query, you can see which needs it by checking the schema documentation on the GraphiQL interface (I'm pushing a change today that includes that detail)

The queries look the same overall, usually only removing the `account` param or something similar and a new header is added to the request: `Authentication: Bearear #{token}`

We need to finish implementing this as @amiskov already did a lot of those changes and decide how we are going to support this new token we need to add to certain queries. Might be a good opportunity to change to the RemoteData pattern if it doesn't take a lot of time

let me know if you need a quick call or something similar too!

Reply:
Alright, I'll take a look, and if I need more details I'll let you know!